058-kraft-upgrades-and-downgrades.md

```diff
@@ -0,0 +1,213 @@
+If the validation fails, the existing metadata version will be used and the error will be raised in log and in the `.status` section of the Kafka CR.
+
+##### Downgrading metadata version
+
+When downgrading the `metadata.version` in Kafka, the operator will always try to do a _safe_ downgrade only.
+And if the safe downgrade fails, it will report an error (in log and in `.status` section of the Kafka CR) and continue reconciliation without changing the metadata version.
+
+In case the safe downgrade is not supported, an _unsafe_ downgrade can be attempted (as explained in one of the earlier sections, the _unsafe_ downgrade is not implemented in current Kafka versions, but it is planned to be supported in the future).
+Unsafe upgrade might result in metadata loss that might cause various problems to the Kafka cluster.
+Users who decide they want to do the unsafe downgrade can do so manually:
+* Pause the reconciliation of the Kafka CR
+* Update the Kafka CR to the updated metadata version (to avoid the operator upgrading it back immediately after the reconciliation is unpaused)
+* Use the Kafka Admin API or CLI tools to do the unsafe downgrade
+* Unpause the reconciliation of the Kafka CR
+
+##### Configuring the initial metadata version
+
+When creating a new Kafka cluster, the initial metadata version can be configured using the `--release-version` option of the `kafka-storage.sh` script.
+To allow users to deploy Kafka clusters with older metadata version, Strimzi will also store the desired `metadata.version` in the per-broker configuration Config Map. 
+The value will be mounted into the Kafka pods and used when formatting the storage of a new Kafka cluster.
+That way - unlike when using environment variable - we will not need to roll the Pod every time the `metadata.version` changes (since in such cases we can update it dynamically without rolling the pods).
+
+#### Upgrades
+
+Upgrade can happen in two situations:
+* After the upgrade of the Strimzi Cluster Operator:
+    * When the new version of the Cluster Operator supports a new default Kafka version and `Kafka.spec.kafka.version` is not set.
+    * When the user changes `Kafka.spec.kafka.version` in parallel with the Strimzi Cluster Operator upgrade.
+* When user requests Kafka upgrade by changing `Kafka.spec.kafka.version`.
+
+In both cases, the Strimzi Cluster Operator will:
+1. Validate that the metadata version currently used is compatible with the new Kafka version (it will check that it is lower than the new Kafka version).
+   This check is expected to pass for upgrades since a Kafka cluster that would not pass it would be invalid already before the upgrade.
+2. Roll all Kafka pods to use the containers with the new Kafka version.
+
+Once all the Kafka pods are rolled, the next step will depend on whether `Kafka.spec.kafka.metadataVersion` is set or not.
+When it is not set, Strimzi will automatically update it to the default version corresponding to the new Kafka version.
```

Comment:
Does 'automatically update it' mean updating the ``Kafka.spec.kafka.metadataVersion`?

Reply:
No, this refers to updating the metadata version in the Kafka cluster itself. It will not update the Kafka CR, so `Kafka.spec.kafka.metadataVersion` will remain unset (and having it unset means the default metadata version for given Kafka version is used, so it will be in-sync).

Reply:
(I also clarified this in the proposal to make it clear there)